src/stack.rs

```diff
@@ -338,25 +354,63 @@ impl<A: AtatClient, T: Timer<TIMER_HZ>, const TIMER_HZ: u32, const TX_SIZE: usiz
 
     /// Asserts that the given socket is connected and returns otherwise the appropriate error
     fn assert_socket_connected(&self, socket: &Socket) -> nb::Result<(), Error> {
-        if self.sockets[socket.link_id] == SocketState::Closing {
+        if self.session.is_socket_closing(socket) {
             return nb::Result::Err(nb::Error::Other(Error::ClosingSocket));
         }
 
-        if self.sockets[socket.link_id] != SocketState::Connected {
+        if !self.session.is_socket_connected(socket) {
             return nb::Result::Err(nb::Error::Other(Error::SocketUnconnected));
         }
 
         nb::Result::Ok(())
     }
+}
+
+impl<const RX_SIZE: usize> Session<RX_SIZE> {
+    /// Fetches the next open socket ID and returns None in case no socket is available
+    fn get_next_open(&self) -> Option<usize> {
+        self.sockets.iter().position(|state| state.state == ConnectionState::Closed)
+    }
+
+    /// Returns true if data is available for the given socket
+    fn is_data_available(&self, socket: &Socket) -> bool {
+        self.sockets[socket.link_id].data_available > 0
```

Comment:
Just an observation: By indexing directly on the array, you might get an out-of-bounds read if the socket doesn't exist.

This can happen if you still have a socket from a previous session. If you use `session.is_data_available(...)` with that socket after recreating the session, you'd get a panic.

In the end, sockets are just plain integers. If the sockets would be tied to the session (session ID?), you could detect this and reject the socket. Alternatively, sockets could be "carried over" to the new session when restarting, as long as this doesn't break any invariants (I'm not sure).

Reply:
> By indexing directly on the array, you might get an out-of-bounds read if the socket doesn't exist

Should not be a problem, since the array is always initialized with the maximum number of possible sockets and no sockets beyond are returned in `socket()` method.

> If the sockets would be tied to the session (session ID?), you could detect this and reject the socket. Alternatively, sockets could be "carried over" to the new session when restarting, as long as this doesn't break any invariants (I'm not sure).

The thought has come to my mind as well, but currently I can't think of any problems if the socket is taken over. The internal state is always up-to-date and calls to send() is rejected as `UconnectedSocket`.

The only question is what should happen when `receive()` is called. Currently, it just returns an empty array. 
The state is intentionally not checked beforehand to allow fetching data even if the other side has closed the connection (closing state).